### PR TITLE
chore: set the namespace if empty for alertmanagers to the current na…

### PR DESCRIPTION
**Pull request by yp969803**
…mespace in Prometheus CR (#1078)

## Description
fixes:  #1078 
Setting the namespace for alertmanagers to the current namespace in the alertsmanager of a Prometheus spec

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [x] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification
make test

## Changelog entry

```
set the namespace if empty for alertmanagers to the current namespace in Prometheus CR
```

### CODE REVIEW

**Comment by yp969803** [comment]
@slashpai  @ArthurSens can you review this pr

**Review by simonpasquier — COMMENTED** [review]
From my testing, the `namespace` field can't be empty because it doesn't have the `omitempty` tag.

```
# prometheuses.monitoring.coreos.com "prometheus" was not valid:                                                                                                                        
# * spec.alerting.alertmanagers[0].namespace: Required value   
```

What could be done though is to make it optional but require a minimal length of 1 if defined. Then if the field is undefined, the operator should inject the Prometheus object's namespace.

```
type AlertmanagerEndpoints struct {
        // Namespace of the Endpoints object.
        // +kubebuilder:validation:MinLength=1
        // +optional
        Namespace *string `json:"namespace"`
        // Name of the Endpoints object in the namespace.
        // +required
        // +kubebuilder:validation:MinLength=1
        Name string `json:"name"`

```

**Comment by yp969803** [comment]
@simonpasquier got it

**Comment by yp969803** [comment]
@simonpasquier  done the changes

**Review by simonpasquier — COMMENTED** [review]
Can you add some test cases to `testPrometheusCRDValidation` in test/e2e/prometheus_test.go?

Also the unit tests fail.

**Comment by yp969803** [comment]
@simonpasquier done the changes, added a test

File: pkg/prometheus/promcfg.go [hunk 1]
```diff
@@ -1767,7 +1780,7 @@ func (cg *ConfigGenerator) generateAlertmanagerConfig(alerting *monitoringv1.Ale
 		// config as well, make sure to path the right namespace here.
 		cfg = addTLStoYaml(cfg, "", am.TLSConfig)
 
-		cfg = append(cfg, cg.generateK8SSDConfig(monitoringv1.NamespaceSelector{}, am.Namespace, apiserverConfig, store, kubernetesSDRoleEndpoint, nil))
+		cfg = append(cfg, cg.generateK8SSDConfig(monitoringv1.NamespaceSelector{}, *am.Namespace, apiserverConfig, store, kubernetesSDRoleEndpoint, nil))
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:1783]: (suggestion) use ptr.Deref() to return a default value if the pointer is nil
```suggestion
		cfg = append(cfg, cg.generateK8SSDConfig(monitoringv1.NamespaceSelector{}, ptr.Deref(am.Namespace, cg.prom.GetObjectMeta().GetName()), apiserverConfig, store, kubernetesSDRoleEndpoint, nil))
```

File: pkg/prometheus/promcfg.go [hunk 2]
```diff
@@ -685,6 +685,19 @@ func (cg *ConfigGenerator) GenerateServerConfiguration(
...
+			if am.Namespace == nil {
+				if promNamespace := cg.prom.GetObjectMeta().GetNamespace(); promNamespace != "" {
+					am.Namespace = &promNamespace
+					continue
+				}
+				am.Namespace = &defaulNamespace
+			}
+		}
+	}
+
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:700]: Defaulting can happen later
```suggestion
```

File: pkg/prometheus/promcfg_test.go [hunk 3]
```diff
@@ -896,12 +896,13 @@ func TestK8SSDConfigGeneration(t *testing.T) {
...
 
 func TestAlertmanagerBearerToken(t *testing.T) {
+	defaultNamespace := "default"
 	p := defaultPrometheus()
 	p.Spec.Alerting = &monitoringv1.AlertingSpec{
 		Alertmanagers: []monitoringv1.AlertmanagerEndpoints{
 			{
 				Name:            "alertmanager-main",
-				Namespace:       "default",
+				Namespace:       &defaultNamespace,
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg_test.go:905]:
```suggestion
				Namespace:       ptr.To("default")
```

File: pkg/prometheus/promcfg_test.go [hunk 4]
```diff
@@ -896,12 +896,13 @@ func TestK8SSDConfigGeneration(t *testing.T) {
 }
 
 func TestAlertmanagerBearerToken(t *testing.T) {
+	defaultNamespace := "default"
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg_test.go:899]:
```suggestion
```

File: pkg/prometheus/promcfg_test.go [hunk 5]
```diff
@@ -965,7 +967,7 @@ func TestAlertmanagerBasicAuth(t *testing.T) {
 					Alertmanagers: []monitoringv1.AlertmanagerEndpoints{
 						{
 							Name:      "alertmanager-main",
-							Namespace: "default",
+							Namespace: &defaultNamespace,
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg_test.go:970]: same here and for all other lines below

File: test/e2e/prometheus_test.go [hunk 6]
```diff
@@ -4628,10 +4657,18 @@ func testPrometheusCRDValidation(t *testing.T) {
 				return
 			}
 
-			_, err := framework.CreatePrometheusAndWaitUntilReady(context.Background(), ns, prom)
+			prom, err := framework.CreatePrometheusAndWaitUntilReady(context.Background(), ns, prom)
 			if err != nil {
 				t.Fatalf("expected no error but got %v", err)
 			}
+			if test.name == "no-namespace" {
```
Comment (simonpasquier) [inline test/e2e/prometheus_test.go:4664]: Hmm we don't need additional testing. We only want to ensure that an object with an empty namespace is accepted by the API.

File: pkg/prometheus/promcfg_test.go
Comment (simonpasquier) [inline pkg/prometheus/promcfg_test.go:1]: We would a new test verifying the generated configuration when namespace is nil.

**Comment by yp969803** [comment]
@simonpasquier done the above changes

**Comment by yp969803** [comment]
@simonpasquier while running my wriiten e2e test, getting the error
```
prometheus_test.go:4720: AlmEndpoint namespace not set to promNs
```
Meaning if the namespace is leaved empty, it is not setting it to prometheus namespace, namespace is nil

**Comment by yp969803** [comment]
this is the snippet where the test is failing
```
p, err := framework.MonClientV1.Prometheuses(ns).Create(context.Background(), prom, metav1.CreateOptions{})
			if err != nil {
				t.Fatalf("expected no error but got %v", err)
			}
			if almanager := p.Spec.Alerting; almanager != nil {
				for _, endpoint := range almanager.Alertmanagers {
					if endpoint.Namespace == nil {
						t.Fatalf("AlmEndpoint namespace not set to promNs")
					} else if *(endpoint.Namespace) != p.Namespace {
						t.Fatalf("AlmEndpoint namespace not match with promNs")
					}
				}
			}
```

File: pkg/prometheus/promcfg.go [hunk 7]
```diff
@@ -685,6 +685,18 @@ func (cg *ConfigGenerator) GenerateServerConfiguration(
...
 	}
 
+	if alerting != nil {
+		for _, am := range alerting.Alertmanagers {
+			if am.Namespace == nil {
+				if promNamespace := cg.prom.GetObjectMeta().GetNamespace(); promNamespace != "" {
+					am.Namespace = &promNamespace
+					continue
+				}
+				am.Namespace = ptr.To("default")
```
[the remark this answers is not in the document]
Comment (yp969803) [reply pkg/prometheus/promcfg.go]: @simonpasquier done the above change

**Comment by yp969803** [comment]
again getting the same error,  
```
TestAllNS/y/PrometheusCRDValidation/no-namespace_Alertmanager_Endpoint
    prometheus_test.go:4720: AlmEndpoint namespace not set to promNs
  ```

**Comment by yp969803** [comment]
@simonpasquier done the above changes

**Comment by yp969803** [comment]
@simonpasquier getting
```
Error: pkg/prometheus/promcfg.go:1815:84: cannot use am.Namespace (variable of type *string) as string value in argument to store.ForNamespace
```
but cant find store.ForNamespace in promcfg.go

**Comment by ArthurSens** [comment]
> simonpasquier getting
> 
> ```
> Error: pkg/prometheus/promcfg.go:1815:84: cannot use am.Namespace (variable of type *string) as string value in argument to store.ForNamespace
> ```
> 
> but cant find store.ForNamespace in promcfg.go

Usually, compilation problems are highlighted by popular IDEs. 

The error message also tells the exactly file and line that this error comes from `pkg/prometheus/promcfg.go:1815` (Line 1815 in pkg/prometheus/promcfg.go file)

**Comment by yp969803** [comment]
@simonpasquier  can u review the pr, i have done the changes

**Comment by yp969803** [comment]
```
if alerting != nil {
		for _, am := range alerting.Alertmanagers {
			if err := am.Validate(); err != nil {
				return nil, fmt.Errorf("generating alerting configuration failed: %w", err)
			}
			if am.Namespace == nil {
				am.Namespace = ptr.To(cg.prom.GetObjectMeta().GetNamespace())

```
dont know why this thing id not called in test @simonpasquier

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 8]
```diff
@@ -1711,7 +1712,10 @@ type APIServerConfig struct {
 // +k8s:openapi-gen=true
 type AlertmanagerEndpoints struct {
 	// Namespace of the Endpoints object.
-	Namespace string `json:"namespace"`
+	// If not set, the object will be discovered in the namespace of the Prometheus object.
+	// +optional
+	Namespace *string `json:"namespace"`
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/prometheus_types.go:1717]:
```suggestion
	Namespace *string `json:"namespace,omitempty"`
```

File: test/framework/prometheus.go [hunk 9]
```diff
@@ -323,7 +323,7 @@ func (f *Framework) AddAlertingToPrometheus(p *monitoringv1.Prometheus, ns, name
 	p.Spec.Alerting = &monitoringv1.AlertingSpec{
 		Alertmanagers: []monitoringv1.AlertmanagerEndpoints{
 			{
-				Namespace: ns,
+				Namespace: &ns,
```
Comment (simonpasquier) [inline test/framework/prometheus.go:326]: (nit)
```suggestion
				Namespace: ptr.To(ns),
```

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 10]
```diff
@@ -1711,7 +1712,10 @@ type APIServerConfig struct {
...
+
 	// Name of the Endpoints object in the namespace.
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/prometheus_types.go:1719]:
```suggestion
	// Name of the Endpoints object in the namespace.
	// +kubebuilder:validation:MinLength:=1
	// +required
```

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 11]
```diff
@@ -1785,6 +1789,17 @@ type AlertmanagerEndpoints struct {
...
+func (e *AlertmanagerEndpoints) Validate() error {
+	if e == nil {
+		return nil
+	}
+	if e.Name == "" {
+		return errors.New("No AlertmanagerEndpoint name specified")
+	}
+	return nil
+}
+
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/prometheus_types.go:1802]: The CRD validation markers make this obsolete.
```suggestion
```

**Comment by yp969803** [comment]
@simonpasquier done the changes, i think now it can be merged :)

**Comment by yp969803** [comment]
now getting 
```
prometheus_test.go:4731: expected no error but got creating 1 Prometheus instances failed (test): [Prometheus.monitoring.coreos.com](http://prometheus.monitoring.coreos.com/) "test" is invalid: spec.alerting.alertmanagers[0].namespace: Required value
```
cant find the code snippet where it is failing

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 12]
```diff
@@ -1711,7 +1711,11 @@ type APIServerConfig struct {
 // +k8s:openapi-gen=true
 type AlertmanagerEndpoints struct {
 	// Namespace of the Endpoints object.
-	Namespace string `json:"namespace"`
+	// If not set, the object will be discovered in the namespace of the Prometheus object.
+	// +kubebuilder:validation:MinLength:=1
+	// +required
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/prometheus_types.go:1716]:
```suggestion
	// +optional
```

Comment (yp969803) [reply pkg/apis/monitoring/v1/prometheus_types.go:1716]: But the issue is with endpoint namespace

Comment (simonpasquier) [reply pkg/apis/monitoring/v1/prometheus_types.go:1716]: The namespace field was "optional" before in the sense that you could set it to an empty value. Making it required would mean a breaking change.

**Comment by yp969803** [comment]
@simonpasquier done the changes, thanku for ur support and guidance.

**Comment by yp969803** [comment]
@ArthurSens @simonpasquier all tests are passed i think now it can be merged :)

File: pkg/apis/monitoring/v1/prometheus_types.go (hunk 8 above)
Comment (simonpasquier) [reply pkg/apis/monitoring/v1/prometheus_types.go:1718]: we still need :arrow_up:

File: pkg/prometheus/promcfg.go [hunk 13]
```diff
@@ -736,6 +736,15 @@ func (cg *ConfigGenerator) GenerateServerConfiguration(
 		return nil, fmt.Errorf("generating storage_settings configuration failed: %w", err)
 	}
 
+	if alerting != nil {
+		for _, am := range alerting.Alertmanagers {
+			if am.Namespace == nil {
+				am.Namespace = ptr.To(cg.prom.GetObjectMeta().GetNamespace())
+
+			}
+		}
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:745]: (nit) can you move this to appendAlertingConfig()?

File: pkg/prometheus/promcfg.go [hunk 14]
```diff
@@ -1823,8 +1832,9 @@ func (cg *ConfigGenerator) generateAlertmanagerConfig(alerting *monitoringv1.Ale
 	}
 
 	alertmanagerConfigs := make([]yaml.MapSlice, 0, len(alerting.Alertmanagers))
-	s := store.ForNamespace(cg.prom.GetObjectMeta().GetNamespace())
 	for i, am := range alerting.Alertmanagers {
+		s := store.ForNamespace(*am.Namespace)
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:1836]: Hmm, it's changing the current behavior which is to look for references into the Prometheus namespace.

**Comment by yp969803** [comment]
@simonpasquier done the changes

**Comment by yp969803** [comment]
@simonpasquier 
```PromAdditionalAlertManagerConfig
    prometheus_test.go:1318: waiting for 1 Prometheus instances timed out (test): prometheus allns-y-promadditionalalertmanagerconfig-sh7i4r-0-1e7d61dd/test failed to become available: client rate limiter Wait returned an error: context deadline exceeded: context deadline exceeded
    ```
    currently getting this, should i increase the context-time of the test or there is somthing other error?

File: pkg/prometheus/promcfg.go [hunk 15]
```diff
@@ -807,6 +807,13 @@ func (cg *ConfigGenerator) appendAlertingConfig(
 		return cfg, nil
 	}
 
+	for _, am := range alerting.Alertmanagers {
+		// Set the namespace of the Alertmanager to the same namespace as the Prometheus resource if not set.
+		if am.Namespace == nil {
+			am.Namespace = ptr.To(cg.prom.GetObjectMeta().GetNamespace())
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:813]: you're modifying a copy, not the original value. If you incorporate the suggestion below, this code change isn't needed anymore.

File: pkg/prometheus/promcfg.go [hunk 16]
```diff
@@ -1848,7 +1857,7 @@ func (cg *ConfigGenerator) generateAlertmanagerConfig(alerting *monitoringv1.Ale
 
 		cfg = addTLStoYaml(cfg, cg.prom.GetObjectMeta().GetNamespace(), am.TLSConfig)
 
-		cfg = append(cfg, cg.generateK8SSDConfig(monitoringv1.NamespaceSelector{}, am.Namespace, apiserverConfig, s, kubernetesSDRoleEndpoint, nil))
+		cfg = append(cfg, cg.generateK8SSDConfig(monitoringv1.NamespaceSelector{}, *am.Namespace, apiserverConfig, s, kubernetesSDRoleEndpoint, nil))
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:1860]: A nil pointer triggers a panic.
```suggestion
		ns := ptr.Deref(am.Namespace, cg.prom.GetObjectMeta().GetNamespace())
		cfg = append(cfg, cg.generateK8SSDConfig(monitoringv1.NamespaceSelector{}, ns, apiserverConfig, s, kubernetesSDRoleEndpoint, nil))
```

**Comment by yp969803** [comment]
@simonpasquier done the changes

**Comment by yp969803** [comment]
@simonpasquier @ArthurSens can it be merger now ?

**Review by simonpasquier — APPROVED** [review]

**Review by simonpasquier — APPROVED** [review]

**Pull request merged**